src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -230,24 +228,19 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
             else if (!Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics)
             {
                 needToCheckConversionToObject = true;
-                objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
-                conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
             }
 
             Debug.Assert(appendCalls.Length == unconvertedInterpolatedString.Parts.Length);
             Debug.Assert(appendCalls.All(a => a is { HasErrors: true } or BoundCall { Arguments: { Length: > 0 } } or BoundDynamicInvocation));
-            int baseStringLength = 0;
-            int numFormatHoles = 0;
 
-            foreach (var currentPart in unconvertedInterpolatedString.Parts)
+            if (needToCheckConversionToObject)
             {
-                if (currentPart is BoundStringInsert insert)
+                TypeSymbol objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                BindingDiagnosticBag conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false); ;
```

Comment:
```suggestion
                BindingDiagnosticBag conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
```

Reply:
Only saw this cause I made the **exact** same mistake today and Chris caught me 

https://github.com/dotnet/roslyn/pull/54675#discussion_r666587369